### ISSUE TITLE
gulp.src ignore option does not match documentation

### ISSUE THREAD

**Issue by breyed** [gulpjs/gulp#1296]
The [gulp API docs](https://github.com/gulpjs/gulp/blob/master/docs/API.md) indicate that gulp supports all options supported by node-glob:

> gulp adds some additional options in addition to the options supported by node-glob and glob-stream:

[Node-glob](https://github.com/isaacs/node-glob#options) says it supports an `ignore` option.

> ignore Add a pattern or an array of patterns to exclude matches.

Furthermore, node-glob says to use `ignore` and not to use negation:

> The intent for negation would be for a pattern starting with `!` to match everything that doesn't match the supplied pattern. However, the implementation is weird. It is better to use the `ignore` option to set a pattern or set of patterns to exclude from matches.

However, as indicated in issue #1176, `ignore` does not work in gulp. Something is broken: the docs, the implementation, or both. Step one is to determine the desired state. Step two is to update the docs and/or code. What should the behavior be?

**Comment by jmm** [comment]
I agree, I don't understand the response in #1176, and the tone seems unnecessarily negative. How would someone submit a documentation PR when no clarification was provided on how the current documentation differs from the expected behavior?

My understanding is that `gulp.src()` or `glob-stream` special-cases the negated patterns, but since it uses `glob` under the hood I'd rather use `opts.ignore` than negated patterns, as advised by that package.

Also, there was a bug in `glob` that prevented `opts.ignore` from working in connection with the way `gulp.src()` or `glob-stream` utilizes `glob`: https://github.com/isaacs/node-glob/issues/194. That was fixed relatively recently. I don't know if there's another issue that prevents `opts.ignore` from working through `gulp.src()`, but it's at least worth checking / updating your `glob` dependency to see if that fixes it.

**Comment by callumacrae** [comment]
A PR fixing the docs would definitely be welcome :)

**Comment by jmm** [comment]
@callumacrae What would constitute fixing them? What about the current documentation is inaccurate?

**Comment by contra** [comment]
We do not pass through the ignore option - this is on purpose. A PR to correct the docs would be appreciated if you want to help out.

**Comment by breyed** [comment]
Some questions to help the documentation update be complete:

* Are there any other options besides `ignore` that gulp doesn't pass through?
* Node-glob warned of negation weirdness. Is there anything in particular that we should caution users against?
* Node-glob warned about dropping negation support? If this happens, will gulpfiles break? Is there anything we should include to assuage concerns or help users future-proof their code?

**Comment by contra** [comment]
@breyed 

- AFAIK just ignore
- globs are executed in order, but this is already in the docs about negation in src
- glob already dropped negation, which doesn't effect us because we never relied on it and never passed through the ignore option

We have always done our own negation for consistency and performance reasons.

Just to explain more:

```js
gulp.src(['client/*.js', '!client/b*.js', 'client/bad.js'])
```

ends up internally (in .src) as

```js
getFiles('client/*.js')
  .pipe(ignoreFiles('client/b*.js'))
  .pipe(getFiles('client/bad.js'))
```

So given the following directory structure

```
client/
  a.js
  bob.js
  bad.js
```

You will get `a.js` and `bad.js` out of the stream.

**Comment by contra** [comment]
```js
gulp.src(['app/**/*', '!app/node_modules/**/*'])
```

Will still traverse the node modules folder, but then negate at the vinyl-fs level which is a huge CPU and memory hog. Somebody sent a PR to fix this by using the node-glob ignore option so it won't crawl anything it shouldn't, but that PR never got merged due to some issues in node-glob: https://github.com/gulpjs/glob-stream/pull/40

**Comment by jmm** [comment]
@contra 
> We do not pass through the ignore option - this is on purpose.
> AFAIK we pass through everything, but ignore just does nothing because we do our own negation

I think you're saying that the gulp stuff doesn't *set* an `ignore` option (it implements negation by other means), but does pass it through if the user passed it. Is there some specific reason at this time that passing `ignore` to `gulp.src()` wouldn't be viable? As of a few months ago there was a bug in `glob` that would prevent `ignore` from working in conjunction with the way the gulp stuff utilizes `glob`, but that was supposed to be fixed as of v5.0.13.


> We have always done our own negation for consistency and performance reasons.
> Will still traverse the node modules folder, but then negate at the vinyl-fs level which is a huge CPU and memory hog.

These comments are kind of hard to reconcile. The custom negation stuff is better or worse for performance? If `glob`'s `ignore` works is there a performance advantage to the custom gulp negation? Not sure what the consistency issues are.

**Comment by contra** [comment]
@jmm 

- What's your use case for not wanting to put the negations in the array? Why are you so keen on using the ignore option?
- As far as "not viable" I don't think the ignore option to the underlying glob module will play well with our stuff, but feel free to try that yourself
- The performance isn't black and white, it has nuances - doing our own negation is fast in most cases, and slow in that one specific case. 

I guess I'm just confused about this whole series of github threads: Can you clearly explain what the desired outcome of this looks like to you?

**Comment by jmm** [comment]
@contra
* 
  > What's your use case for not wanting to put the negations in the array? Why are you so keen on using the ignore option?

  Originally, because the gulp documentation implies that it passes the patterns through to node-glob:

  > `glob` refers to [node-glob syntax](https://github.com/isaacs/node-glob)

  Where the node-glob documentation linked to tells you not to use negated patterns and to use `ignore` instead:

  > The intent for negation would be for a pattern starting with `!` to match everything that *doesn't* match the supplied pattern. However, the implementation is weird, and for the time being, this should be avoided. The behavior is deprecated in version 5, and will be removed entirely in version 6.

  > Note: In version 5 of this module, negation and comments are disabled by default....They are going away entirely in version 6.

  > It is better to use the `ignore` option

  I don't see anything in the gulp docs that explains that it special-cases negated patterns. In fact negated patterns or ignoring paths are barely even *mentioned*. I see one example of a negated pattern, with no commentary, in the main [README](https://github.com/gulpjs/gulp/blob/master/README.md). I don't see *anything* about it in the [`gulp.src()` docs](https://github.com/gulpjs/gulp/blob/master/docs/API.md#gulpsrcglobs-options). Realistically, most people aren't going to think there's any reason to look at the glob-stream docs, which in any case still don't explain that negated patterns are special-cased and still say that options are passed through to node-glob.
  
  I only discovered that gulp special-cases negated patterns when I noticed that the `ignore` option didn't work. I fully expected it to since the gulp docs say that the options are:

  > [passed] to node-glob through glob-stream.

  That's when I discovered https://github.com/isaacs/node-glob/issues/194 and chalked it up to that. 

  Now that I know all of this I'd still prefer to be consistent with the underlying implementation (node-glob) and use `ignore` unless there's an important reason to rely on gulp's special-casing of the negated patterns.

* 
  > As far as "not viable" I don't think the ignore option to the underlying glob module will play well with our stuff, but feel free to try that yourself

  Yeah I'll plan to give it a shot.

> I guess I'm just confused about this whole series of github threads: Can you clearly explain what the desired outcome of this looks like to you?

What series is that, this and #1176? Desired outcome:

* Find out if `ignore` does or doesn't work and why.
* Find out if there is or isn't some important reason to rely on gulp's special-casing of negated patterns instead.
* Make any necessary corrections to the documenation.
* Don't have the documentation say you can't use `ignore` unless you really can't and shouldn't.
* Help you understand how the documentation and comments regarding this are confusing for your users.

**Comment by contra** [comment]
@jmm

People should not be using the ignore option, this is an API decision we made and why the option is not documented (or recommended, or talked about) anywhere - however, we do not prevent people from passing the ignore option to node-glob under the hood if they are familiar with the internals and have a special case.

Nowhere in our documentation do we tell people to use the ignore option, and we have a section about globs and glob negation that shows how we expect users to do it. If a user wants to know how to do glob negation they can read our documentation that explains how to do glob negation.

I can't understand why you're so heated about this tiny documentation change - I am sincerely sorry if our documentation confused you. Really. But let's focus on fixing it so we don't have to keep sending paragraphs back and forth.

Can you send a PR that solves this? I can't seem to get on your wavelength and understand your desired solution, so I think it would be easier for both of us if you go ahead and make your changes.

**Comment by breyed** [comment]
@jmm 
> Nowhere in our documentation do we tell people to use the ignore option

The old documentation effectively did tell people to use `ignore`. When the docs just say "this module passes options through to module *x*", you're telling users to freely use all the options of module *x*.

A PR to fix the documentation already exists. What @contra is asking about is whether we should go further. He's asking whether the API decision to not use `ignore` is the right decision. A PR isn't a bad way to answer the question: see if it's practical to implement an API that matches node-glob's.

The one caveat I'd put out there is that if anyone would be against an `ignore`-based API even if it has a solid implementation behind it, please speak now before @contra invests time into the PR.

**Comment by jmm** [comment]
@breyed 
> The old documentation effectively did tell people to use `ignore`.

You mean the *current* documentation, right? :smiley:

> What @contra is asking about is whether we should go further. He's asking whether the API decision to not use `ignore` is the right decision.

I don't get that impression.

> A PR to fix the documentation already exists.

Yes, I know, I've seen it. :smiley: We might have a different idea of what "fix" means. That's what I was referring to in this comment:

> Don't have the documentation say you can't use `ignore` unless you really can't and shouldn't.



> A PR isn't a bad way to answer the question: see if it's practical to implement an API that matches node-glob's.

An API that matches node-glob's would just be:

```js
gulp.src(['positive', 'globs'], {
  ignore: ['negative', 'globs'],
});
```

I think you're referring to the PR that was mentioned earlier, https://github.com/gulpjs/glob-stream/pull/40. My understanding is that that relates to continuing to accept negated patterns in `gulp.src()` and then under the hood send them to node-glob via `ignore` (?). If that would work, I don't know what the point would be, and in any case it would mean I can do what I want, which is just pass `ignore` as an option to `gulp.src()` and have it work.

FYI I'm not ignoring your post @ contra -- I'll be replying.

**Comment by contra** [comment]
I think the current PR to fix the documentation solves the problem

**Comment by jmm** [comment]
Well clearly my comments are going to be disregarded, but I'll reply anyway for posterity.

> Nowhere in our documentation do we tell people to use the ignore option

Not explicitly, no. But it's implied, as I and others have explained.

> we have a section about globs and glob negation that shows how we expect users to do it. If a user wants to know how to do glob negation they can read our documentation that explains how to do glob negation.

Ok, where is that? It's not at [`gulp.src()`](https://github.com/gulpjs/gulp/blob/47623606afb698f66a4085ad6f73bc7270ad1654/docs/API.md#gulpsrcglobs-options). Following the links there indicates to use `options.ignore` and *not* negated globs.

> But let's focus on fixing it so we don't have to keep sending paragraphs back and forth.
> 
> Can you send a PR that solves this? I can't seem to get on your wavelength and understand your desired solution, so I think it would be easier for both of us if you go ahead and make your changes.

I would rather use `ignore` than negated globs (as the documentation currently indicates can be done). So if that works I'm all set and there's no PR necessary. I assumed that was the case when https://github.com/isaacs/node-glob/issues/194 got fixed. But now you're going to make the documentation explicitly say that `gulp.src()` *doesn't* support `ignore`, so even if it does work following isaacs/node-glob#194 (I haven't had the opportunity to try it yet) I'll have to use negated globs or just do my own globbing.

**Comment by breyed** [comment]
@contra Perhaps I don't understand the process used by the gulp team, but I don't understand why you closed this issue while the [API docs in master](https://github.com/gulpjs/gulp/blob/master/docs/API.md) are still unchanged. Wouldn't you want to merge the PR before closing?

**Comment by breyed** [comment]
> Well clearly my comments are going to be disregarded...

@jmm I understand what would lead you to feel this way, but it's not right to accuse people of things they might do but haven't actually done. For example, I don't know why @contra cut off discussion on #1176. It seems a harsh use of the tool, but perhaps there is a good reason that just hasn't been stated (in which case, it should be). But even if @contra made a mistake of not listening well to the community in the past (I don't claim to know one way or the other whether he has), there's nothing stopping him from learning from that mistake and being a great community leader going forward.

Given the bugs and API changes spanning multiple components and versions, this is a complex topic. We all need to be careful to listen well to each other. We also need as best we can to take a step back, and look at the situation from the perspective of a user just learning gulp who knows nothing about its internals. They are the ones who benefit most from a great API and great documentation.

**Comment by breyed** [comment]
For clarity, we could split the discussion into two issues. We could limit this thread to ensuring documentation matches the gulp-specific negation-based implementation. @jmm, you may want to open a new issue titled something like "Use an ignore option as the primary means of excluding files". The issue would discuss and possibly implement supporting `ignore` and presenting it as the primary API for excluding files (leaving `!` for backward compatibility). I like how such a change would increase consistency with the glob API, but don't know enough to lead such a change.

**Comment by jmm** [comment]
@breyed

> but it's not right to accuse people of things they might do but haven't actually done.

That's a great sentiment, but I don't know why you're directing that comment at me. Before I could reply to comments by @ contra that were directed to me, a resolution, that largely disregards my comments on the subject so far, was announced and the issue was closed. Those things were actually done before my comment that you quoted.

**Comment by breyed** [comment]
@jmm You're points about the past are valid, but the statement I took issue with was about the future. Closed issues can be reopened, and we all have ample opportunity to hear each other out, regardless of what has happened in the past.

**Comment by jmm** [comment]
@breyed
> Closed issues can be reopened

Yup, in the vein of "anything's possible".

> and we all have ample opportunity to hear each other out

You're free to interpret anything however you want. I was actively participating in the discussion, and even explicitly said I was going to reply. The only clearer signal I could ask for that there's no interest in hearing me out than pronouncing an outcome and closing the issue would be locking it.

**Comment by phated** [comment]
The PR hasn't been merged yet. I think we need to keep this open until that PR gets resolved.

**Comment by contra** [comment]
Look, I don't know what I did to personally ruin your life @jmm but you need to stop this. Every time I respond you go line by line and tear apart every word I'm saying for no reason. I'm not here to fight anyone, I don't understand where your aggression is coming from. I'm sorry if you think I'm a dickhead based on some github issue that happened a year ago, I was probably having a shitty day. I don't know how many times I need to apologize before you give up your vendetta.

Honestly? I closed this issue because I don't want to deal with your needless aggression anymore and the problem is already being fixed. I don't have time to keep writing 5 paragraph responses defending my choice of wording on issues that were posted a year ago. This discussion seems like it's never going to end, no matter what I do to try to fix the situation. You just want to argue forever for no apparent reason.

I explained the reasoning behind not using the ignore option as the primary means of negation here https://github.com/gulpjs/gulp/issues/1296#issuecomment-144583691 - there is no reason to open a new issue raising the question again.

**Comment by breyed** [comment]
@contra, "ruin life", "tear apart", "aggression", "think I'm a...": none of these are substantiated, and even if they were, would be far nearer to personal attacks than to helping improve gulp. I really think the disconnect stemmed from two helpful people who are passionate about gulp, but with perspectives based on experience in one case more on the internals and in another more on the external usage. I don't feel you need to bow out of the discussion. Just please be patient in your responses if something that is obvious to you isn't obvious to others.

Case in point (and bringing us back to the actual topic at hand), it seems that in the comment you linked to, one of features you were alluding to is that  `!` negation lets you interleave includes and excludes, which lets you use the ordering to limit the scope of excludes. By contrast, an `ignore` negation always applies to all included globs. Perhaps you're saying that order-based exclusion is a powerful enough feature that we don't need a separate thread to rule out eliminating it. If so, I tend to agree. (I must admit I didn't pick up on the significance of order-based exclusion from the linked comment alone.)

I tend to wonder two things:
* How useful would an `ignore` option be so that people who happen to have array of exclusions don't have to make a `map` call to add `!`s to them? I haven't run into that scenario - not sure if anyone has.
* If order-based exclusion is useful, why doesn't node-glob offer it in a straightforward way?

**Comment by jmm** [comment]
I can't tell you how to feel, but even though I've experienced a fair amount of frustration with the way this discussion has unfolded, my comments have been on topic and I object to characterizing them as an "aggressive vendetta". I hope anyone reading those comments reads the whole thread -- if it's not edited it speaks for itself.

I also have better uses of my time than to keep reiterating some of the same questions and concerns. The more comments there are that I can't understand or don't jibe with other information I have, like the docs, the more I'm going to have to question them to advance the discussion. If you say you're confused and I don't know how or why you're confused, I'm going to try to explain thoroughly enough to clear up the confusion. When referring to external information (like docs or other issues), or previous comments, it can be very helpful to link to and or / quote from them, so I frequently do that. In several cases you reference external things in a non-specific way that makes it impossible for me to know what you're referring to. In several cases you say things that are inconsistent with previous comments or other information like your own documentation. All of these things contribute to more back and forth.

Ordinarily I would quote and respond to other things from @ contra's latest post, but won't in this case for obvious reasons.

Future expectations revised:
> The only clearer signal I could ask for that there's no interest in hearing me out than pronouncing an outcome and closing the issue would be locking it.

Were you having a bad day when you wrote your latest post? Do you need someone to talk to? Though I imagine you'd rather talk to your friends, I'm serious, if you need it I will find some time for a private conversation about anything you want someone to listen about, except this issue. (You can email me at `contra-2015-10 {@} jessemccarthy {DOT} net`. I'll delete the address in a month.) Just because, but let me also say explicitly that I appreciate your work on these projects. If you weren't having a bad day and that's just how you roll, then of course, do you.

**Comment by phated** [comment]
Going through some glob-stream issues, I've found some issues and PRs that are relevant to this discussion:

* https://github.com/gulpjs/glob-stream/issues/24 - agrees that glob negation is a bottleneck in gulp and suggests using the ignore option in node-glob which resulted in...
* https://github.com/gulpjs/glob-stream/pull/40 - seems to switch to the ignore option from node-glob but is blocked by...
* https://github.com/isaacs/node-glob/pull/166 - which seems to have been ignored for quite some time

**Comment by douxing** [comment]
i solve the ignore problem using another tricky way:
1. using [glob.sync](https://www.npmjs.com/package/glob#glob-sync-pattern-options) with ignore option in it
2. pass the result to the gulp.src as the first parameter

```javascript
var result = glob.sync ... // choose your files the glob way
gulp.src result ... // using the files
```
